File: 3sem/4lab/Banks/BankAccounts/Factories/DepositBankAccountFactory.cs
using Banks.Clocks;
using Banks.Entities;
using Banks.Exceptions;

namespace Banks.BankAccounts.Factories;

public class DepositBankAccountFactory : IBankAccountFactory
{
    public DepositBankAccountFactory(decimal deposit)
    {
        if (deposit < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        Deposit = deposit;
    }

    public decimal Deposit { get; }

    public IBankAccount Create(Bank bank, Client client, IClock clock)
    {
        return new DepositBankAccount(Deposit, bank, client, clock);
    }
}

File: 4sem/1lab/src/main/java/banksRemake/domain/bankAccounts/Factories/DepositBankAccountFactory.java
package banksRemake.domain.bankAccounts.Factories;

import banksRemake.domain.bankAccounts.BankAccount;
import banksRemake.domain.bankAccounts.DepositBankAccount;
import banksRemake.domain.clocks.Clock;
import banksRemake.domain.entities.Bank;
import banksRemake.domain.entities.Client;
import banksRemake.domain.exceptions.InvalidBankAccountOperationException;
import lombok.Getter;

/**
 * Factory creating deposit bank accounts.
 */
public class DepositBankAccountFactory implements BankAccountFactory {
    @Getter
    private final double deposit;

    public DepositBankAccountFactory(double deposit) {
        this.deposit = deposit;
    }

    @Override
    public BankAccount create(Bank bank, Client client, Clock clock) throws InvalidBankAccountOperationException {
        return new DepositBankAccount(deposit, bank, client, clock);
    }
}
